### PR TITLE
VB-638 HMPPS Domain Events

### PR DESCRIPTION
**Pull request by philip-milne**
## What does this pull request do?
Adds domain events for Booked and Cancelled visits with localstack test container.
Updates Circle for AWS requirements.
Update docker to allow containers and tests to run at the same time,

## What is the intent behind these changes?
Adds and SNS service with unit tests, for Nomis sync.

Note:
local dev env and CI config out-of-scope.
Send Exception disabled pending VB-671.

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -17,7 +17,7 @@ services:
       - PRISON_API_URL=http://prison-api:8080
```
Comment (AledWEvans) [inline docker-compose.yml:17]: I wonder if we should have a docker-compose-local-dev.yml and just this one...

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/config/VisitSchedulerExceptionHandler.kt [hunk 2]
```diff
@@ -74,7 +75,7 @@ class VisitSchedulerExceptionHandler {
 
   @ExceptionHandler(MissingServletRequestParameterException::class)
   fun handleValidationException(e: MissingServletRequestParameterException): ResponseEntity<ErrorResponse> {
-    log.debug("Bad Request (400) returned", e.message)
+    log.debug("Bad Request (400) returned {}", e.message)
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/config/VisitSchedulerExceptionHandler.kt:78]: $e.message can we do that in Kotlin?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/config/VisitSchedulerExceptionHandler.kt [hunk 3]
```diff
@@ -170,6 +171,20 @@ class VisitSchedulerExceptionHandler {
...
+  @ExceptionHandler(PublishEventException::class)
+  fun handlePublishEventException(e: PublishEventException): ResponseEntity<ErrorResponse?>? {
+    log.debug("Publish event exception caught: {}", e.message)
+    return ResponseEntity
+      .status(HttpStatus.INTERNAL_SERVER_ERROR)
+      .body(
+        ErrorResponse(
+          status = HttpStatus.INTERNAL_SERVER_ERROR,
+          userMessage = "Failed to publish event: ${e.cause?.message}",
+          developerMessage = e.message
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/config/VisitSchedulerExceptionHandler.kt:183]: should exclude this if production or perhaps if production exception class name

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SnsService.kt [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    val log: Logger = LoggerFactory.getLogger(this::class.java)
+  }
+}
+
+data class AdditionalInformation(
+  val reference: String,
+)
+
+@Suppress("unused")
+data class HMPPSDomainEvent(
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SnsService.kt:98]: DTO, do you like internal classes?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SnsService.kt [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    const val EVENT_PRISON_VISIT_BOOKED = "prison-visit.booked"
+    const val EVENT_PRISON_VISIT_BOOKED_DESC = "Prison Visit Booked"
+    const val EVENT_PRISON_VISIT_CANCELLED = "prison-visit.cancelled"
+    const val EVENT_PRISON_VISIT_CANCELLED_DESC = "Prison Visit Cancelled"
+
+    val log: Logger = LoggerFactory.getLogger(this::class.java)
+  }
+}
+
+data class AdditionalInformation(
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SnsService.kt:93]: DTO

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SnsService.kt [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    eventType: String,
+    description: String,
+    occurredAt: Instant,
+    prisonerId: String,
+    additionalInformation: AdditionalInformation
+  ) : this(
+    eventType,
+    1,
+    description,
+    occurredAt.toOffsetDateFormat(),
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SnsService.kt:116]: Nice, time zones...

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SnsService.kt [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+  val version: Int,
+  val description: String,
+  val occurredAt: String,
+  val prisonerId: String,
+  val additionalInformation: AdditionalInformation,
+) {
+  constructor(
+    eventType: String,
+    description: String,
+    occurredAt: Instant,
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SnsService.kt:109]: If you need LocalDateTime Or  LocalDate just insists on it!

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SnsService.kt [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+  }
+
+  fun sendVisitCancelledEvent(visit: VisitDto) {
+    log.debug("Sending visit cancelled event")
+    publishToDomainEventsTopic(
+      HMPPSDomainEvent(
+        eventType = EVENT_PRISON_VISIT_CANCELLED,
+        version = EVENT_PRISON_VISIT_VERSION,
+        description = EVENT_PRISON_VISIT_CANCELLED_DESC,
+        occurredAt = Instant.now().toString(),
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SnsService.kt:51]:
```
        occurredAt = LocalDateTime.now().toString()
Or
        occurredAt = LocalDate.now().toString()
```

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SnsService.kt [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+        prisonerId = visit.prisonerId,
+        additionalInformation = AdditionalInformation(
+          reference = visit.reference
+        )
+      )
+    )
+  }
+
+  private fun publishToDomainEventsTopic(payload: HMPPSDomainEvent) {
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SnsService.kt:60]: payload,.. I would rather event, payload is a little generic

File: src/test/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/integration/container/LocalStackContainer.kt [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+      setWaitStrategy(Wait.forListeningPort())
+      start()
+      followOutput(logConsumer)
+    }
+  }
+
+  private fun isLocalStackRunning(): Boolean =
+    try {
+      val serverSocket = ServerSocket(4566)
+      serverSocket.localPort == 0
```
Comment (AledWEvans) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/integration/container/LocalStackContainer.kt:36]: I wonder if we could do this a different way?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/integration/container/PostgresContainer.kt [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+      withReuse(true)
+      start()
+      followOutput(logConsumer)
+    }
+  }
+
+  private fun isPostgresRunning(): Boolean =
+    try {
+      val serverSocket = ServerSocket(5432)
+      serverSocket.localPort == 0
```
Comment (AledWEvans) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/integration/container/PostgresContainer.kt:35]: as above?

**Review by AledWEvans — COMMENTED** [review]
Just some points, will talk to discuss

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/config/VisitSchedulerExceptionHandler.kt (hunk 2 above)
Comment (philip-milne) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/config/VisitSchedulerExceptionHandler.kt:78]: Sure we would have to update all of the debug for this class, this change is just to put in the missing syntax.  We can create a ticket to review the debug for the whole app

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/config/VisitSchedulerExceptionHandler.kt (hunk 3 above)
Comment (philip-milne) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/config/VisitSchedulerExceptionHandler.kt:183]: Same again, need a ticket to review the error handling

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SnsService.kt (hunk 4 above)
Comment (philip-milne) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SnsService.kt:98]: Ill review the encapsulation, like you say its a serialized object we might want to expose it.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SnsService.kt (hunk 6 above)
Comment (philip-milne) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SnsService.kt:116]: This need a good test something i wanted a second pair of eyes on when it comes to roundtrip serialisation and object mapping constructors.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SnsService.kt (hunk 7 above)
Comment (philip-milne) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SnsService.kt:109]: Im re-using the signatures from the rest of the MoJ projects - Instant was not my first choice

File: src/test/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/integration/container/LocalStackContainer.kt (hunk 10 above)
Comment (philip-milne) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/integration/container/LocalStackContainer.kt:36]: Again using the existing MoJ projects for consistency - could externalise or randomise.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SnsService.kt (hunk 8 above)
Comment (philip-milne) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SnsService.kt:51]: Yeah, although it works and matches the status quo, i'm not overly happy with this code.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/config/VisitSchedulerExceptionHandler.kt (hunk 2 above)
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/config/VisitSchedulerExceptionHandler.kt:78]: another time... no probs

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/config/VisitSchedulerExceptionHandler.kt (hunk 3 above)
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/config/VisitSchedulerExceptionHandler.kt:183]: sure

**Review by AledWEvans — APPROVED** [review]

**Pull request merged**